xarray/test/__init__.py

```diff
@@ -154,67 +155,91 @@ def assertWarns(self, message):
+        assert_xarray_identical(d1, d2)
 
     def assertDatasetAllClose(self, d1, d2, rtol=1e-05, atol=1e-08):
-        self.assertEqual(sorted(d1, key=str), sorted(d2, key=str))
-        self.assertItemsEqual(d1.coords, d2.coords)
-        for k in d1:
-            v1 = d1.variables[k]
-            v2 = d2.variables[k]
-            self.assertVariableAllClose(v1, v2, rtol=rtol, atol=atol)
+        assert_xarray_allclose(d1, d2, rtol=rtol, atol=atol)
 
     def assertCoordinatesEqual(self, d1, d2):
-        self.assertEqual(sorted(d1.coords), sorted(d2.coords))
-        for k in d1.coords:
-            v1 = d1.coords[k]
-            v2 = d2.coords[k]
-            self.assertVariableEqual(v1, v2)
+        assert_xarray_equal(d1, d2)
 
     def assertDataArrayEqual(self, ar1, ar2):
-        self.assertVariableEqual(ar1, ar2)
-        self.assertCoordinatesEqual(ar1, ar2)
+        assert_xarray_equal(ar1, ar2)
 
     def assertDataArrayIdentical(self, ar1, ar2):
-        self.assertEqual(ar1.name, ar2.name)
-        self.assertDatasetIdentical(ar1._to_temp_dataset(),
-                                    ar2._to_temp_dataset())
+        assert_xarray_identical(ar1, ar2)
 
     def assertDataArrayAllClose(self, ar1, ar2, rtol=1e-05, atol=1e-08):
-        self.assertVariableAllClose(ar1, ar2, rtol=rtol, atol=atol)
-        self.assertCoordinatesEqual(ar1, ar2)
-
+        assert_xarray_allclose(ar1, ar2, rtol=rtol, atol=atol)
+
+def assert_xarray_equal(a, b):
+    import xarray as xr
+    ___tracebackhide__ = True  # noqa: F841
+    assert type(a) == type(b)
+    if isinstance(a, (xr.Variable, xr.DataArray, xr.Dataset)):
+        assert a.equals(b), '{}\n{}'.format(a, b)
+    elif isinstance(a, xr.core.coordinates.AbstractCoordinates):
+        assert sorted(a) == sorted(b)
+        [assert_xarray_equal(a[k], b[k]) for k in a]
+    else:
+        raise TypeError('{} not supported by assertion comparison'
+                        .format(type(a)))
+
+def assert_xarray_identical(a, b):
+    import xarray as xr
+    ___tracebackhide__ = True  # noqa: F841
+    assert type(a) == type(b)
+    if isinstance(a, xr.DataArray):
+        assert a.name == b.name
+        assert_xarray_identical(a._to_temp_dataset(), b._to_temp_dataset())
+    elif isinstance(a, (xr.Dataset, xr.Variable)):
+        assert a.identical(b), '{}\n{}'.format(a, b)
+    else:
+        raise TypeError('{} not supported by assertion comparison'
+                        .format(type(a)))
+
+def assert_xarray_allclose(a, b, rtol=1e-05, atol=1e-08):
+    import xarray as xr
+    ___tracebackhide__ = True  # noqa: F841
+    assert type(a) == type(b)
+    if isinstance(a, xr.DataArray):
+        assert_xarray_allclose(a.variable, b.variable, rtol=rtol, atol=atol)
+        assert_xarray_equal(a.coords, b.coords)
+    elif isinstance(a, xr.Dataset):
+        assert sorted(a, key=str) == sorted(a, key=str)
+        assert_xarray_equal(a.coords, b.coords)
```

Comment:
I think this is why many (not all) of the new tests are failing (at least the unicode/bytes ones, because we have a hack that says unicode/bytes can still be "close" if they encode the same string). Previously we used `allclose` for all variables, now `coords` are required to be equal.

I can see why this makes sense, but maybe better to keep with what we have now or add a keyword argument to switch.